tensorflow/tensorflow-engine/src/main/java/ai/djl/tensorflow/engine/TfOpBuilder.java

```diff
@@ -0,0 +1,190 @@
+
+import static org.tensorflow.internal.c_api.global.tensorflow.*;
+
+import ai.djl.Device;
+import ai.djl.engine.EngineException;
+import ai.djl.ndarray.NDArray;
+import ai.djl.ndarray.types.DataType;
+import ai.djl.util.Preconditions;
+import java.nio.charset.StandardCharsets;
+import java.util.Arrays;
+import org.bytedeco.javacpp.BytePointer;
+import org.bytedeco.javacpp.IntPointer;
+import org.bytedeco.javacpp.PointerPointer;
+import org.bytedeco.javacpp.PointerScope;
+import org.tensorflow.internal.c_api.TFE_Context;
+import org.tensorflow.internal.c_api.TFE_Op;
+import org.tensorflow.internal.c_api.TFE_TensorHandle;
+import org.tensorflow.internal.c_api.TF_Status;
+
+final class TfOpBuilder {
+
+    /** This value should be >= to the maximum number of outputs in any op */
+    private static final int MAX_OUTPUTS_PER_OP = 1000;
+
+    private TfNDManager manager;
+    private TFE_Op opHandle;
+
+    @SuppressWarnings({"unchecked", "try"})
+    TfOpBuilder(TfNDManager manager, TFE_Context eagerSessionHandle, String operation) {
+        this.manager = manager;
+        try (PointerScope scope = new PointerScope()) {
+            TF_Status status = TF_Status.newStatus();
+            opHandle = TFE_Op.newOp(eagerSessionHandle, operation, status);
+            status.throwExceptionIfNotOK();
+            // keep the native pointer alive outside of the scope
+            opHandle.retainReference();
+        }
+    }
+
+    // please make sure you close the output manually or attach to NDManager
+    @SuppressWarnings({"unchecked", "try"})
+    private TFE_TensorHandle[] execute(TFE_Op opHandle) {
+        try (PointerScope scope = new PointerScope()) {
+            IntPointer numReturnValues = new IntPointer(1).put(MAX_OUTPUTS_PER_OP);
+            PointerPointer<TFE_TensorHandle> returnValues =
+                    new PointerPointer<>(MAX_OUTPUTS_PER_OP);
+            TF_Status status = TF_Status.newStatus();
+            TFE_Execute(opHandle, returnValues, numReturnValues, status);
+            status.throwExceptionIfNotOK();
+
+            TFE_TensorHandle[] results = new TFE_TensorHandle[numReturnValues.get()];
+            for (int i = 0; i < results.length; ++i) {
+                results[i] =
+                        returnValues
+                                .get(TFE_TensorHandle.class, i)
+                                .withDeallocator()
+                                .retainReference();
+            }
+            return results;
+        }
+    }
+
+    public NDArray[] build() {
+        TFE_TensorHandle[] handles = execute(opHandle);
+        NDArray[] outputs = new NDArray[handles.length];
+        for (int i = 0; i < handles.length; ++i) {
+            // attach the TfNDArray along with pointer to manager
+            outputs[i] = new TfNDArray(manager, handles[i]);
+        }
+        // close the opHandle
+        opHandle.close();
+        return outputs;
+    }
+
+    public NDArray buildSingletonOrThrow() {
```

Comment:
Can we just return NDList, and user can call NDList.gingleOrThrow()

Reply:
that's what i did in the first place but realize we will create lots of unused NDList